src/syscalls/interfaces/uhyve.rs

```diff
@@ -215,78 +216,41 @@ impl SyscallInterface for Uhyve {
-					syscmdsize.argsz[i] as usize * mem::size_of::<*const u8>(),
-					1,
-				);
-				argv_phy[i] =
-					paging::virtual_to_physical(VirtAddr(argv[i] as u64)).as_u64() as *const u8;
-			}
+		// determine the number of arguments and environment variables
+		let mut syscmdsize = SysCmdsize::new();
+		uhyve_send(UHYVE_PORT_CMDSIZE, &mut syscmdsize);
+
+		// create array to receive all arguments
+		let mut argv = Box::new(vec![ptr::null(); syscmdsize.argc as usize]);
+		let mut argv_phy = vec![ptr::null(); syscmdsize.argc as usize];
+		for i in 0..syscmdsize.argc as usize {
+			argv[i] = crate::__sys_malloc(syscmdsize.argsz[i] as usize * mem::size_of::<u8>(), 1);
+			argv_phy[i] =
+				paging::virtual_to_physical(VirtAddr(argv[i] as u64)).as_u64() as *const u8;
+		}
 
-			// create array to receive the environment
-			let env_raw = crate::__sys_malloc(
-				(syscmdsize.envc + 1) as usize * mem::size_of::<*const u8>(),
-				mem::size_of::<*const u8>(),
-			) as *mut *const u8;
-			let env_phy_raw = crate::__sys_malloc(
-				(syscmdsize.envc + 1) as usize * mem::size_of::<*const u8>(),
-				mem::size_of::<*const u8>(),
-			) as *mut *const u8;
-			let env = slice::from_raw_parts_mut(env_raw, (syscmdsize.envc + 1) as usize);
-			let env_phy = slice::from_raw_parts_mut(env_phy_raw, (syscmdsize.envc + 1) as usize);
-			for i in 0..syscmdsize.envc as usize {
-				env[i] = crate::__sys_malloc(
-					syscmdsize.envsz[i] as usize * mem::size_of::<*const u8>(),
-					1,
-				);
-				env_phy[i] =
-					paging::virtual_to_physical(VirtAddr(env[i] as u64)).as_u64() as *const u8;
-			}
-			env[syscmdsize.envc as usize] = ptr::null_mut();
-			env_phy[syscmdsize.envc as usize] = ptr::null_mut();
-
-			// ask uhyve for the environment
-			let mut syscmdval =
-				SysCmdval::new(VirtAddr(argv_phy_raw as u64), VirtAddr(env_phy_raw as u64));
-			uhyve_send(UHYVE_PORT_CMDVAL, &mut syscmdval);
-
-			// free temporary array
-			crate::__sys_free(
-				argv_phy_raw as *mut u8,
-				syscmdsize.argc as usize * mem::size_of::<*const u8>(),
-				mem::size_of::<*const u8>(),
-			);
-			crate::__sys_free(
-				env_phy_raw as *mut u8,
-				(syscmdsize.envc + 1) as usize * mem::size_of::<*const u8>(),
-				mem::size_of::<*const u8>(),
-			);
-
-			(
-				syscmdsize.argc,
-				argv_raw as *const *const u8,
-				env_raw as *const *const u8,
-			)
+		// create array to receive the environment
+		let mut env = Box::new(vec![ptr::null(); syscmdsize.envc as usize + 1]);
+		let mut env_phy = vec![ptr::null(); syscmdsize.envc as usize + 1];
+		for i in 0..syscmdsize.envc as usize {
+			env[i] = crate::__sys_malloc(syscmdsize.envsz[i] as usize * mem::size_of::<u8>(), 1);
+			env_phy[i] = paging::virtual_to_physical(VirtAddr(env[i] as u64)).as_u64() as *const u8;
 		}
+
+		// ask uhyve for the environment
+		let mut syscmdval = SysCmdval::new(
+			VirtAddr(argv_phy.as_ptr() as u64),
+			VirtAddr(env_phy.as_ptr() as u64),
+		);
+		uhyve_send(UHYVE_PORT_CMDVAL, &mut syscmdval);
```

Comment:
I guess this is a blocking send and it is ok that `argv_phy` and `env_phy` are destructed when we return from this function? Is `syscmdval` modified or why is it passed as `&mut`?

Reply:
The hypervisor is able to modify the record. The compiler should be aware that the data can be modified.

Reply:
So it is more like a send_receive function?

Reply:
Yes, it is a hypercall to the VM.

Reply:
Ah, ok. I think the naming is not perfect in that case. Also I am wondering if &mut is fine in case the hypervisor can modify the data, should it not be marked as volatile? But let's discuss this in slack 😉 , I'll approve this PR.